src/core/lib/transport/interception_chain.cc

```diff
@@ -0,0 +1,160 @@
+  }
+
+  void StartCall(UnstartedCallHandler unstarted_call_handler) override {
+    destination_->HandleCall(unstarted_call_handler.StartCall(stack_));
+  }
+
+ private:
+  RefCountedPtr<CallFilters::Stack> stack_;
+  RefCountedPtr<CallDestination> destination_;
+};
+
+class TerminalInterceptor final : public UnstartedCallDestination {
+ public:
+  explicit TerminalInterceptor(
+      RefCountedPtr<CallFilters::Stack> stack,
+      RefCountedPtr<UnstartedCallDestination> destination)
+      : stack_(std::move(stack)), destination_(std::move(destination)) {}
+
+  void Orphaned() override {
+    stack_.reset();
+    destination_.reset();
+  }
+
+  void StartCall(UnstartedCallHandler unstarted_call_handler) override {
+    unstarted_call_handler.SpawnGuarded(
+        "start_call",
+        Map(interception_chain_detail::HijackCall(unstarted_call_handler,
+                                                  destination_, stack_),
+            [](ValueOrFailure<HijackedCall> hijacked_call) -> StatusFlag {
+              if (!hijacked_call.ok()) return Failure{};
+              ForwardCall(hijacked_call.value().original_call_handler(),
+                          hijacked_call.value().MakeLastCall());
+              return Success{};
+            }));
+  }
+
+ private:
+  RefCountedPtr<CallFilters::Stack> stack_;
+  RefCountedPtr<UnstartedCallDestination> destination_;
+};
+}  // namespace
+
+///////////////////////////////////////////////////////////////////////////////
+// InterceptionChain::Builder
+
+void InterceptionChainBuilder::AddInterceptor(
+    absl::StatusOr<RefCountedPtr<Interceptor>> maybe_interceptor) {
+  if (!status_.ok()) return;
+  if (!maybe_interceptor.ok()) {
+    status_ = maybe_interceptor.status();
+    return;
+  }
+  auto interceptor = std::move(maybe_interceptor.value());
+  interceptor->filter_stack_ = MakeFilterStack();
+  if (top_interceptor_ == nullptr) {
+    top_interceptor_ = std::move(interceptor);
+  } else {
+    Interceptor* previous = top_interceptor_.get();
+    while (previous->wrapped_destination_ != nullptr) {
+      previous = DownCast<Interceptor*>(previous->wrapped_destination_.get());
+    }
+    previous->wrapped_destination_ = std::move(interceptor);
+  }
+}
+
+absl::StatusOr<RefCountedPtr<UnstartedCallDestination>>
+InterceptionChainBuilder::Build(FinalDestination final_destination) {
+  if (!status_.ok()) return status_;
+  // Build the final UnstartedCallDestination in the chain - what we do here
+  // depends on both the type of the final destination and the filters we have
+  // that haven't been captured into an Interceptor yet.
+  absl::StatusOr<RefCountedPtr<UnstartedCallDestination>> terminator = Match(
+      final_destination,
+      [this](RefCountedPtr<UnstartedCallDestination> final_destination)
+          -> absl::StatusOr<RefCountedPtr<UnstartedCallDestination>> {
+        if (stack_builder_.has_value()) {
+          // TODO(ctiller): consider interjecting a hijacker here
```

Comment:
I don't quite follow this comment.  Why would we want a hijacker here?

Reply:
unclear now... possible a remnant from an earlier iteration (this took a bunch of them)